File: sub-quote-java/src/main/java/top/x07/tstClient/traceCompress/DataFlags.java
package top.x07.tstClient.traceCompress;

import java.nio.ByteBuffer;

public class DataFlags extends javolution.io.Struct {
	private final Unsigned8 f_low = new Unsigned8(2);  // b-> lastClose
	private final Unsigned8 f_high = new Unsigned8(2);  // b-> lastClose
	private final Unsigned8 f_open = new Unsigned8(2);  // b-> lastClose
	private final Unsigned8 f_lastClose = new Unsigned8(2);  // b-> 0

	private final Unsigned8 f_bp2 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_bp1 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_volume = new Unsigned8(2);  // b-> 0
	private final Unsigned8 f_close = new Unsigned8(2);  // b-> lastClose

	private final Unsigned8 f_bv1 = new Unsigned8(2);  // b-> 0
	private final Unsigned8 f_bp5 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_bp4 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_bp3 = new Unsigned8(2);  // b-> close

	private final Unsigned8 f_bv5 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_bv4 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_bv3 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_bv2 = new Unsigned8(2);  // b-> bv1

	private final Unsigned8 f_sp4 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_sp3 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_sp2 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_sp1 = new Unsigned8(2);  // b-> close

	private final Unsigned8 f_sv3 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_sv2 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_sv1 = new Unsigned8(2);  // b-> bv1 [special!!]
	private final Unsigned8 f_sp5 = new Unsigned8(2);  // b-> close

	private final Unsigned8 f_limitUp = new Unsigned8(2);  // b-> lastClose
	private final Unsigned8 f_neipan = new Unsigned8(2);  // b-> volume
	private final Unsigned8 f_sv5 = new Unsigned8(2);  // b-> bv1
	private final Unsigned8 f_sv4 = new Unsigned8(2);  // b-> bv1

	private final Unsigned8 f_temp2 = new Unsigned8(2);  // b-> close
	private final Unsigned8 f_avPrice = new Unsigned8(2);   // 期货的结算价（平均价） b-> 0
	private final Unsigned8 f_position = new Unsigned8(2);  // 期货持仓 b-> 0
	private final Unsigned8 f_limitDown = new Unsigned8(2);  // b-> lastClose

	public DataFlags(byte[] b) {
		this.setByteBuffer(ByteBuffer.wrap(b), 0);
	}

	public short lastClose() { return f_lastClose.get(); }
	public short open() { return f_open.get(); }
	public short high() { return f_high.get(); }
	public short low() { return f_low.get(); }
	public short close() { return f_close.get(); }
	public short volume() { return f_volume.get(); }

	public short bp1() { return f_bp1.get(); }
	public short bp2() { return f_bp2.get(); }
	public short bp3() { return f_bp3.get(); }
	public short bp4() { return f_bp4.get(); }
	public short bp5() { return f_bp5.get(); }
	public short bv1() { return f_bv1.get(); }
	public short bv2() { return f_bv2.get(); }
	public short bv3() { return f_bv3.get(); }
	public short bv4() { return f_bv4.get(); }
	public short bv5() { return f_bv5.get(); }

	public short sp1() { return f_sp1.get(); }
	public short sp2() { return f_sp2.get(); }
	public short sp3() { return f_sp3.get(); }
	public short sp4() { return f_sp4.get(); }
	public short sp5() { return f_sp5.get(); }
	public short sv1() { return f_sv1.get(); }
	public short sv2() { return f_sv2.get(); }
	public short sv3() { return f_sv3.get(); }
	public short sv4() { return f_sv4.get(); }
	public short sv5() { return f_sv5.get(); }

	public short neipan() { return f_neipan.get(); }
	public short limitUp() { return f_limitUp.get(); }
	public short limitDown() { return f_limitDown.get(); }
	public short position() { return f_position.get(); }

	public short avPrice() { return f_avPrice.get(); }
	public short temp2() { return f_temp2.get(); }
}

File: sub-quote-py/DataFlags.py
import struct

class DataFlags(object):
    f_lastClose = 0  # b-> 0
    f_open = 2  # b-> lastClose
    f_high = 4  # b-> lastClose
    f_low = 6  # b-> lastClose

    f_close = 0  # b-> lastClose
    f_volume = 2  # b-> 0
    f_bp1 = 4  # b-> close
    f_bp2 = 6  # b-> close

    f_bp3 = 0  # b-> close
    f_bp4 = 2  # b-> close
    f_bp5 = 4  # b-> close
    f_bv1 = 6  # b-> 0

    f_bv2 = 0  # b-> bv1
    f_bv3 = 2  # b-> bv1
    f_bv4 = 4  # b-> bv1
    f_bv5 = 6  # b-> bv1

    f_sp1 = 0  # b-> close
    f_sp2 = 2  # b-> close
    f_sp3 = 4  # b-> close
    f_sp4 = 6  # b-> close

    f_sp5 = 0  # b-> close
    f_sv1 = 2  # b-> bv1 [special!!]
    f_sv2 = 4  # b-> bv1
    f_sv3 = 6  # b-> bv1

    f_sv4 = 0  # b-> bv1
    f_sv5 = 2  # b-> bv1
    f_neipan = 4  # b-> volume
    f_limitUp = 6  # b-> lastClose

    f_limitDown = 0  # b-> lastClose
    f_position = 2  # 期货持仓 b-> 0
    f_avPrice = 4   # 期货的结算价（平均价） b-> 0
    f_temp2 = 6  # b-> close

    def __init__(self, b):
        u0, u1, u2, u3, u4, u5, u6, u7 = struct.unpack('<BBBBBBBB', b)

        self.lastClose = (u0 >> self.f_lastClose) & 0b11 # b-> 0
        self.open = (u0 >> self.f_open) & 0b11  # b-> lastClose
        self.high = (u0 >> self.f_high) & 0b11  # b-> lastClose
        self.low = (u0 >> self.f_low) & 0b11  # b-> lastClose

        self.close = (u1 >> self.f_close) & 0b11  # b-> lastClose
        self.volume = (u1 >> self.f_volume) & 0b11  # b-> 0
        self.bp1 = (u1 >> self.f_bp1) & 0b11  # b-> close
        self.bp2 = (u1 >> self.f_bp2) & 0b11  # b-> close

        self.bp3 = (u2 >> self.f_bp3) & 0b11  # b-> close
        self.bp4 = (u2 >> self.f_bp4) & 0b11  # b-> close
        self.bp5 = (u2 >> self.f_bp5) & 0b11  # b-> close
        self.bv1 = (u2 >> self.f_bv1) & 0b11  # b-> 0

        self.bv2 = (u3 >> self.f_bv2) & 0b11  # b-> bv1
        self.bv3 = (u3 >> self.f_bv3) & 0b11  # b-> bv1
        self.bv4 = (u3 >> self.f_bv4) & 0b11  # b-> bv1
        self.bv5 = (u3 >> self.f_bv5) & 0b11  # b-> bv1

        self.sp1 = (u4 >> self.f_sp1) & 0b11  # b-> close
        self.sp2 = (u4 >> self.f_sp2) & 0b11  # b-> close
        self.sp3 = (u4 >> self.f_sp3) & 0b11  # b-> close
        self.sp4 = (u4 >> self.f_sp4) & 0b11  # b-> close

        self.sp5 = (u5 >> self.f_sp5) & 0b11  # b-> close
        self.sv1 = (u5 >> self.f_sv1) & 0b11  # b-> bv1 [special!!]
        self.sv2 = (u5 >> self.f_sv2) & 0b11  # b-> bv1
        self.sv3 = (u5 >> self.f_sv3) & 0b11  # b-> bv1

        self.sv4 = (u6 >> self.f_sv4) & 0b11  # b-> bv1
        self.sv5 = (u6 >> self.f_sv5) & 0b11  # b-> bv1
        self.neipan = (u6 >> self.f_neipan) & 0b11  # b-> volume
        self.limitUp = (u6 >> self.f_limitUp) & 0b11  # b-> lastClose

        self.limitDown = (u7 >> self.f_limitDown) & 0b11  # b-> lastClose
        self.position = (u7 >> self.f_position) & 0b11  # 期货持仓 b-> 0
        self.avPrice = (u7 >> self.f_avPrice) & 0b11   # 期货的结算价（平均价） b-> 0
        self.temp2 = (u7 >> self.f_temp2) & 0b11  # b-> close
